openshift/assisted-service
# MGMT-13888: Remove spoke node on BMH delete

This PR builds on https://github.com/openshift/assisted-service/pull/4948 by cleaning and removing the node from the spoke cluster when the BMH is annotated and deleted.

This allows ZTP users (specifically ones using git-ops where they don't have the ability to alter/remove the agent CR) to deprovision nodes using the same mechanisms they used to originally provision them (the BMH).

This does not handle cordoning and draining the spoke node before removing it. That will be handled in a follow up (https://issues.redhat.com/browse/MGMT-13890)

This requires the following fixes in BMO:
https://github.com/metal3-io/baremetal-operator/pull/1184
https://github.com/metal3-io/baremetal-operator/pull/1230

This means that this will only work with 4.13 nightlies after this merges.

## How it works

- When the user annotates the BMH and the converged flow is in use a finalizer is added to the BMH.
- When the BMH is deleted this finalizer (and the work in ~~https://github.com/metal3-io/baremetal-operator/pull/1230~~ https://github.com/metal3-io/baremetal-operator/pull/1236) allows BMAC time to reattach and deprovision the BMH. Once these edits are made BMAC removes our finalizer.
- BMO then boots the host back into the preprovisioning image which runs the ironic agent and cleans the host (this is required to ensure that the host doesn't continue running OCP if/when it is rebooted after it is removed).
- Throughout this process the agent controller has been waiting for the BMH to be fully deleted as a part of its finalizer logic.
- After BMO finishes deprovisioning the host it removes its finalizer on the BMH and allows it to be garbage collected.
- Once the BMH is gone the agent controller contacts the spoke cluster and removes the node record.

## List all the issues related to this PR

https://issues.redhat.com/browse/MGMT-13888
https://issues.redhat.com/browse/RFE-3431


## What environments does this code impact?


## How was this code tested?

Tested manually on the dev-scripts cluster with an edited BMO.

1. Install a spoke using ZTP
2. Annotate the BMHs with `bmac.agent-install.openshift.io/remove-agent-and-node-on-delete: true`
3. Delete a BMH
4. Observe BMH deprovision and node removal




## Checklist

- [ ] This change does not require a documentation update (docstring, `docs`, README, etc) **TODO**

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md
